File: api/models/assignment.py
from sqlalchemy import Column, Integer, DateTime, Date, Time, ForeignKey, Enum, func
from sqlalchemy.orm import relationship
from typing import List
from .base import Base

class Assignment(Base):
    """Model for task assignments to teacher aides."""
    
    __tablename__ = 'assignments'

    id = Column(Integer, primary_key=True)
    task_id = Column(Integer, ForeignKey('tasks.id', ondelete='CASCADE'), nullable=False)
    aide_id = Column(Integer, ForeignKey('teacher_aide.id', ondelete='SET NULL'))
    date = Column(Date, nullable=False)
    start_time = Column(Time, nullable=False)
    end_time = Column(Time, nullable=False)
    status = Column(
        Enum('UNASSIGNED', 'ASSIGNED', 'IN_PROGRESS', 'COMPLETE'),
        nullable=False,
        default='UNASSIGNED'
    )
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, onupdate=func.now())

    # Relationships
    task = relationship("Task", back_populates="assignments")
    aide = relationship("TeacherAide", back_populates="assignments")

    def check_conflicts(self, session) -> List['Assignment']:
        """Check for scheduling conflicts with other assignments."""
        if not self.aide_id:
            return []
        return session.query(Assignment).filter(
            Assignment.aide_id == self.aide_id,
            Assignment.date == self.date,
            Assignment.id != self.id,
            (
                (Assignment.start_time <= self.start_time) & (self.start_time < Assignment.end_time) |
                (Assignment.start_time < self.end_time) & (self.end_time <= Assignment.end_time) |
                (self.start_time <= Assignment.start_time) & (Assignment.start_time < self.end_time)
            )
        ).all()

    def to_dict(self):
        """Convert assignment to dictionary."""
        return {
            'id': self.id,
            'task_id': self.task_id,
            'aide_id': self.aide_id,
            'date': self.date.isoformat(),
            'start_time': self.start_time.isoformat(),
            'end_time': self.end_time.isoformat(),
            'status': self.status,
            'created_at': self.created_at.isoformat(),
            'updated_at': self.updated_at.isoformat()
        } 

File: frontend/src/types/assignment.ts
export interface Assignment {
  id: number;
  task_id: number;
  task_title: string;
  task_category: string;
  start_time: string;
  end_time: string;
  date: string;
  aide_id: number | null;
  aide_name: string | null;
  classroom_id: number | null;
  classroom_name: string | null;
  status: 'UNASSIGNED' | 'ASSIGNED' | 'IN_PROGRESS' | 'COMPLETE';
  is_flexible: boolean;
}

export interface Task {
  id: number;
  title: string;
  category: string;
  start_time: string | null;
  end_time: string | null;
  is_flexible: boolean;
  classroom_id: number | null;
  recurrence_rule: string | null;
}

export interface TeacherAide {
  id: number;
  name: string;
  email: string;
  status: 'ACTIVE' | 'INACTIVE';
}

export interface Classroom {
  id: number;
  name: string;
  capacity: number;
} 